apps/ex_marketer/lib/ex_marketer/crawler/parse.ex

```diff
@@ -0,0 +1,65 @@
+defmodule ExMarketer.Crawler.Parse do
+  def new(body) do
+    body |> parse_document |> filter
+  end
+
+  defp parse_document(body) do
+    {:ok, parsed_document} = body |> Floki.parse_document()
+
+    parsed_document
+  end
+
+  defp filter(parsed_document) do
+    ads_on_top_links = parsed_document |> ads_on_top_links
+    non_ads_links = parsed_document |> non_ads_links
+
+    %{
+      ads_on_top: ads_on_top_links |> Enum.count(),
+      ads_on_top_link: ads_on_top_links,
+      total_ads: parsed_document |> total_ads,
+      total_non_ads: non_ads_links |> Enum.count(),
+      non_ads_link: non_ads_links,
+      total_link: parsed_document |> total_links
+    }
```

Comment:
This approach is valid (I did something similar in my first attempt 😉 ) but lacks a functional programming style which is to passed data to transform it i.e. passing a map `%{}` and using the pipe operation `%{ current | new_key: "new value" }`

Check how Ros did in this PR: https://github.com/rosle/google_crawler/pull/9/files#diff-8c6bc9e0496badfa7b2565b3fdeadd8fR16

This style is more prevalent as it allows to compose piping of operating when using operators like `with`. Basically send the output of one operation to the next. It reads betters and it's easy to add another operation 💭

Reply:
haha, I will check it @olivierobert 👍